src/math/spherical_harmonics.cc

```diff
@@ -181,55 +176,215 @@ namespace rascal {
+                raising_plm_factor *
+                    assoc_legendre_polynom(angular_l, m_count + 1)};
+            // harmonics values
+            harmonics_derivatives(0, l_block_index + angular_l + m_count) =
+                assoc_legendre_polynom(angular_l, m_count) *
+                cos_sin_m_phi(m_count, 0);
+            harmonics_derivatives(0, l_block_index + angular_l - m_count) =
+                assoc_legendre_polynom(angular_l, m_count) *
+                cos_sin_m_phi(m_count, 1);
+
+            double phi_derivative_factor{};
+            if (sin_theta > 0.1) {
+              // singularity at the poles
+              phi_derivative_factor =
+                  m_count / sin_theta *
+                  assoc_legendre_polynom(angular_l, m_count);
+            } else {
+              // singularity at the equator
+              phi_derivative_factor =
+                  -0.5 / cos_theta *
+                  (lowering_plm_factor *
+                       assoc_legendre_polynom(angular_l, m_count - 1) +
+                   raising_plm_factor *
+                       assoc_legendre_polynom(angular_l, m_count + 1));
+            }
+            // d/dx
+            harmonics_derivatives(1, l_block_index + angular_l + m_count) =
+                sin_phi * phi_derivative_factor * cos_sin_m_phi(m_count, 1) +
+                (-0.5 * cos_theta * cos_phi * cos_sin_m_phi(m_count, 0) *
+                 legendre_polynom_difference);
+            harmonics_derivatives(1, l_block_index + angular_l - m_count) =
+                -1.0 * sin_phi * phi_derivative_factor *
+                    cos_sin_m_phi(m_count, 0) +
+                (-0.5 * cos_theta * cos_phi * cos_sin_m_phi(m_count, 1) *
+                 legendre_polynom_difference);
+            // d/dy
+            harmonics_derivatives(2, l_block_index + angular_l + m_count) =
+                -1.0 * cos_phi * phi_derivative_factor *
+                    cos_sin_m_phi(m_count, 1) +
+                (-0.5 * cos_theta * sin_phi * cos_sin_m_phi(m_count, 0) *
+                 legendre_polynom_difference);
+            harmonics_derivatives(2, l_block_index + angular_l - m_count) =
+                cos_phi * phi_derivative_factor * cos_sin_m_phi(m_count, 0) +
+                (-0.5 * cos_theta * sin_phi * cos_sin_m_phi(m_count, 1) *
+                 legendre_polynom_difference);
+            // d/dz
+            harmonics_derivatives(3, l_block_index + angular_l + m_count) =
+                0.5 * sin_theta * cos_sin_m_phi(m_count, 0) *
+                legendre_polynom_difference;
+            harmonics_derivatives(3, l_block_index + angular_l - m_count) =
+                0.5 * sin_theta * cos_sin_m_phi(m_count, 1) *
+                legendre_polynom_difference;
+          }  // if (m_count == 0)
+        }    // for (m_count in [0, l])
+        l_block_index += (2 * angular_l + 1);
+      }  // for (l in [0, lmax])
+      return harmonics_derivatives;
+    }
```

Comment:
What's `d/di`? That sounds like it would be confusing; the way it's done here is how all the other code iterates over arrays with this "compact" structure.